File: components/src/commonMain/kotlin/com/zero/components/layouts/RowComponent.kt
package com.zero.components.layouts

import androidx.compose.foundation.layout.Row
import androidx.compose.runtime.Composable
import com.arkivanov.decompose.extensions.compose.jetbrains.subscribeAsState
import com.zero.common.component.Component
import com.zero.components.ComponentComposer
import com.zero.models.layouts.RowModel

@Composable
fun RowComponent(component: Component, componentModelRetriever: (String) -> Component?) {
    val modelState = component.state.subscribeAsState()
    val model = modelState.value.componentModel as? RowModel

    model?.let {
        Row {
            for (componentId in model.childComponents) {
                componentModelRetriever(componentId)?.let { component ->
                    ComponentComposer(component, componentModelRetriever)
                }
            }
        }
    }
}

File: ios/src/iosAppMain/apple/composers/layouts/RowComponent.swift
//
// Created by Gabriel Timotei on 2/21/22.
//

import SwiftUI
import common

public struct RowComponent: View {
    @ObservedObject
    private var componentState: ObservableValue<ComponentState>
    private let componentModelRetriever: (String) -> Component?

    public init(component: Component, componentModelRetriever: @escaping (String) -> Component?) {
        self.componentModelRetriever = componentModelRetriever

        componentState = ObservableValue<ComponentState>(component.state)
    }

    public var body : some View {
        if let rowModel = componentState.value.componentModel as? RowModel {
            HStack {
                ForEach(rowModel.childComponents, id: \.self) { componentId in
                    ComponentComposer(id: componentId, componentModelRetriever: componentModelRetriever)
                }
            }
        } else {
            EmptyView()
        }
    }
}
